### PR TITLE
Refactored prompt.py to reduce token usage

### PR DESCRIPTION
**Pull request by temotskipa**
Refactored prompt.py to reduce token usage. Also included a band-aid fix to resolve an issue Devin encountered while editing prompt.py. The issue in question was that random example commands included in prompt.py were getting executed in the terminal, which resulted in the agent getting stuck in a loop trying to edit the file unsuccessfully. This issue should be resolved in a timely manner with a proper fix imo.

### CODE REVIEW

**Review by enyst — CHANGES_REQUESTED** [review]
Can you please tell, on what version did you encounter getting stuck in a loop? There is a fix merged on main just these days for that kind of issue, although I suspect it may still be missing commands because of different command ids. Do you by any chance have logs from that issue?

IMHO some edits are a great idea, but I hope @xingyaoww can take a look when we consider this kind of changes.

**Comment by temotskipa** [comment]
I encountered it on the opendevin:main docker image.

**Review by yufansong — CHANGES_REQUESTED** [review]
I prefe to not change the prompts only if we can do experiments to show it do works.

**Comment by neubig** [comment]
I agree with @yufansong , this seems like a great change if it doesn't reduce our accuracy, but I am worried that it might. Maybe we could run swe bench lite with this

**Review by rbren — DISMISSED** [review]
CC @xingyaoww 

Most likely a lot of these extra tokens are required for quality--the LLM often needs some reminding.

But there are also a few typo fixes etc here so we might want to take some of it.

**Comment by temotskipa** [comment]
Is there anything else that I should do or do I just wait for #1941 to be merged?

**Comment by xingyaoww** [comment]
@temotskipa I think that PR is merged :) feel free to re-adjust the prompt!

**Comment by temotskipa** [comment]
I mean I personally think this prompt is fine, but IDK if the maintainers also think so.

**Comment by xingyaoww** [comment]
@temotskipa Can you resolve those merge conflict so we can review again and try to merge it?

**Comment by rbren** [comment]
@temotskipa are you interested in pushing this one forward? @xingyaoww do you have specific changes you'd like to see?

**Comment by yufansong** [comment]
> @temotskipa are you interested in pushing this one forward? @xingyaoww do you have specific changes you'd like to see?

I suggest to hold any prompts change before we finishing all benchmark evaluation.

**Comment by xingyaoww** [comment]
I am mainly looking for fixing the conflicts - once those are fixed, I'm happy with merging this PR (after 2 days - when we finish running all the eval for the paper as yufan suggested)

**Comment by temotskipa** [comment]
So am I supposed to revert the changes and just fix any typos and things like that? It's unclear to me what is conflicting here.

**Comment by xingyaoww** [comment]
@temotskipa Sorry for getting back late! We were running for a huge deadline :(
Feel free to keep whatever you want to change! As long as the conflict is gone, we can review and merge it :)

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:67]:
```suggestion
    numbers = list(range(1, 11))
    return str(numbers)
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:223]:
```suggestion
USER:
Observation:
[1] 126
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:227]:
```suggestion
ASSISTANT:
The server is running on port 5000 with PID 126. You can access the list of numbers in a table format by visiting http://127.0.0.1:5000. Let me know if you have any further requests!
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:168]:
```suggestion
ASSISTANT:
Sure! Let me stop the server first:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:156]:
```suggestion
USER: Now browse the newly started server's homepage and show me the content.

ASSISTANT:
Sure! Let me browse the server's homepage at http://127.0.0.1:5000:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:166]:
```suggestion
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:169]:
```suggestion
The content of the server's homepage is "[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]". If you have any further questions, feel free to ask!

```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:138]:
```suggestion
[1] 124

```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:202]:
```suggestion
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:131]:
```suggestion
ASSISTANT:
I'll rerun the file:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:108]:
```suggestion
ASSISTANT:
It seems that Flask is not installed. Let me install Flask for you:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:100]:
```suggestion
</execute_bash>

```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:96]:
```suggestion
[1] 121[1]+  Exit 1                  python3 app.py > server.log 2>&1

```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:89]:
```suggestion
ASSISTANT:
Now, I'll run the file:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:228]:
```suggestion
ASSISTANT:
Running the updated file:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:160]:
```suggestion
The server is running on port 5000 with PID 124. You can access the list of numbers by visiting http://127.0.0.1:5000. If you have any further questions, feel free to ask!
```

**Review by xingyaoww — APPROVED** [review]
I tweaked the prompt a bit and included most simplifications in the system prompt from @temotskipa -- feel free to review again @enyst @rbren @yufansong

cc @li-boxuan @frankxu2004, with browsing agent delegation, now it seems we need to do this to simply *browse* a page - we start a whole new agent just to execute a `goto` action, anyway we could simplify this?
```
<execute_browse>
Get the content on "http://127.0.0.1:5000"
</execute_browse>
```

**Comment by frankxu2004** [comment]
@xingyaoww we could probably use legacy BrowseURLAction for this. Just an URL in this block?

**Comment by li-boxuan** [comment]
> with browsing agent delegation, now it seems we need to do this to simply browse a page - we start a whole new agent just to execute a goto action, anyway we could simplify this?

We could tweak the prompt more and let CodeActAgent decide whether to delegate. Now I just delegate everything related to browsing activities.

**Comment by xingyaoww** [comment]
@li-boxuan sounds good! Shall we add the legacy action back? Or I'm actually thinking of integrating those browser actions to `agentskills` once we can put browser into the sandbox - and keep the delegation as is for now (per discussion with frank: https://opendevin.slack.com/archives/C06TZ1VCCAD/p1717748986732919?thread_ts=1717748260.194329&cid=C06TZ1VCCAD)

**Comment by neubig** [comment]
Just on a high-level, I think doing the simple thing for now to get this PR finished would be good! But it seems that tests are failing, @xingyaoww if you could take a look that'd be good!

**Review by yufansong — APPROVED** [review]

File: agenthub/codeact_agent/prompt.py
Comment (li-boxuan) [inline agenthub/codeact_agent/prompt.py:94]: @xingyaoww I thought this example would help and encourage the LLM to think step by step, and would be nice to keep it as it is?

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:95]:
```suggestion
I have created a Python file `app.py` that will display a list of numbers from 1 to 10 when you run it. Let me run the Python file for you:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [inline agenthub/codeact_agent/prompt.py:141]:
```suggestion
Now that Flask is installed, let me run the Python file again:
```

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [reply agenthub/codeact_agent/prompt.py]: Good point! Added back!

**Comment by xingyaoww** [comment]
Integration test should be fixed as well -- @rbren @enyst feel free to take a look and unblock the PR if possible :)

**Review by li-boxuan — APPROVED** [review]

**Review by enyst — APPROVED** [review]

**Pull request merged**
